celery/app/task.py

```diff
@@ -882,8 +882,14 @@ def replace(self, sig):
             )
 
         if self.request.chain:
+            options = sig.options.copy()
+            sig.freeze(self.request.id)
             for t in reversed(self.request.chain):
                 sig |= signature(t, app=self.app)
+            sig.set(
+                link=options.get('link', None),
+                link_error=options.get('link_error', None)
+            )
```

Comment:
We shouldn't need to do this for `link_error` because unchaining which occurs when we compose chains with the vertical bar operator duplicates `link_error` errbacks onto all of a chain's encapsulated tasks. We do need to ensure the `link`ed callback is retained (the issue originally identified) but pushing it to the end of the newly reconstructed chain means it'll be run at the wrong time and with the wrong argument(s). This should be roughly:

```suggestion
            sig.freeze(self.request.id)
            if "link" in sig.options:
              sig.tasks[-1].options.setdefault("link", []).append(sig.options["link"])
            for t in reversed(self.request.chain):
                sig |= signature(t, app=self.app)
```

Reply:
The new diff makes this change and there are a couple of new tests to hit the cases I mentioned in previous comments.